flytekit/core/workflow.py

```diff
@@ -277,87 +339,319 @@ def _local_execute(self, ctx: FlyteContext, **kwargs) -> Union[Tuple[Promise], P
+            self._interface = transform_interface_to_typed_interface(self._python_interface)
+
+    def ready(self) -> bool:
+        """
+        This function returns whether or not the workflow is in a ready state, which means
+          * Has at least one node
+          * All workflow inputs are bound
+
+        These conditions assume that all nodes and workflow i/o changes were done with the functions above, which
+        do additional checking.
+        """
+        if len(self.compilation_state.nodes) == 0:
+            return False
+
+        if len(self._unbound_inputs) > 0:
+            return False
+
+        return True
+
+
+class PythonFunctionWorkflow(WorkflowBase, ClassStorageTaskResolver):
+    """
+    More comments to come.
+    """
+
+    def __init__(
+        self,
+        workflow_function: Callable,
+        metadata: Optional[WorkflowMetadata],
+        default_metadata: Optional[WorkflowMetadataDefaults],
+    ):
+        name = f"{workflow_function.__module__}.{workflow_function.__name__}"
+        self._workflow_function = workflow_function
+        native_interface = transform_signature_to_interface(inspect.signature(workflow_function))
+
+        # TODO do we need this - can this not be in launchplan only?
+        #    This can be in launch plan only, but is here only so that we don't have to re-evaluate. Or
+        #    we can re-evaluate.
+        self._input_parameters = None
+        super().__init__(
+            name=name,
+            workflow_metadata=metadata,
+            workflow_metadata_defaults=default_metadata,
+            python_interface=native_interface,
+        )
+
+    @property
+    def function(self):
+        return self._workflow_function
+
+    def task_name(self, t: PythonAutoContainerTask) -> str:
+        return f"{self.name}.{t.__module__}.{t.name}"
+
+    def compile(self, **kwargs):
+        """
+        Supply static Python native values in the kwargs if you want them to be used in the compilation. This mimics
+        a 'closure' in the traditional sense of the word.
+        """
+        ctx = FlyteContext.current_context()
+        self._input_parameters = transform_inputs_to_parameters(ctx, self.python_interface)
+        all_nodes = []
+        prefix = f"{ctx.compilation_state.prefix}-{self.short_name}-" if ctx.compilation_state is not None else None
+        with ctx.new_compilation_context(prefix=prefix, task_resolver=self) as comp_ctx:
+            # Construct the default input promise bindings, but then override with the provided inputs, if any
+            input_kwargs = construct_input_promises([k for k in self.interface.inputs.keys()])
+            input_kwargs.update(kwargs)
+            workflow_outputs = self._workflow_function(**input_kwargs)
+            all_nodes.extend(comp_ctx.compilation_state.nodes)
+
+            # This needs a longer comment.
```

Comment:
is this a todo or meant to be part of this PR?

Reply:
not part of this PR, but i may as well add it.